src/devices/src/legacy/rtc_pl031.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{convert::TryInto, sync::Arc};
+
+use logger::{warn, IncMetric, RTCDeviceMetrics, METRICS};
+
+use crate::BusDevice;
+
+pub type RTCDevice = vm_superio::RTC<Arc<RTCDeviceMetrics>>;
+
+// Implements Bus functions for AMBA PL031 RTC device
+#[cfg(target_arch = "aarch64")]
+impl BusDevice for RTCDevice {
+    fn read(&mut self, offset: u64, data: &mut [u8]) {
+        if data.len() == 4 {
+            // read() function from RTC implementation expects a slice of
+            // len 4, and we just validated that this is the data lengt
+            self.read(offset as u16, data.try_into().unwrap())
```

Comment:
This cast can truncate the data. Should we use a try_from and log an error if the data does not fit?

Reply:
If the guest sends an invalid offset or even an invalid data, it means that the driver is misbehaving. In this case, what is expected from the emulation is "undefined behavior" (i.e. in the emulation we choose what to do, the driver cannot have any expectations of what happens next). For this reason, it might be better to just not add additional checks and keep a cleaner code.

Reply:
The reason is also that these values are checked again when calling read.

Reply:
> If the guest sends an invalid offset or even an invalid data, it means that the driver is misbehaving. In this case, what is expected from the emulation is "undefined behavior" (i.e. in the emulation we choose what to do, the driver cannot have any expectations of what happens next). For this reason, it might be better to just not add additional checks and keep a cleaner code.

I understand that. The logs are however for Firecracker users, and it could help with debugging some obscure cases.
I agree though that the code is a little cleaner this way

Reply:
> The reason is also that these values are checked again when calling read.

When calling read on the RTC from superio? The upstream read works on u16 so the truncation check cannot be done there.

Reply:
What I meant with the check is something else :laughing: I'll try to explain it better.

So let's say that instead of an u16, the driver sends you an u64. We're truncating it, ending up with a u16, which is not going to represent any meaningful value. It's just going to be garbage. This garbage value cannot do anything bad tho because only "good" offsets are allowed into the emulation. That's what I mean with the value being checked in vm-superio.

We should also be mindful about logging too many things that come from untrusted users even if we rate limit the log file.